Nvidia Riva support

kfatehi opened the pull request:
This PR adds support for Nvidia Riva. The Riva stack serves a gRPC service, which seemed non-trivial (impossible?) to interface with from the extension environment, so in this implementation I am relying on a companion [web service](https://github.com/kfatehi/riva_tts_proxy) that conforms an HTTP GET request similar to that used for IBM Watson for riva in order to return the desired ogg file.

# Audio Samples

[riva-english-female-1.webm](https://user-images.githubusercontent.com/175305/235287855-e7cefbcd-496b-47de-b014-e4e2e5596784.webm)

[riva-english-male-1.webm](https://user-images.githubusercontent.com/175305/235287884-8ff41351-d79d-4788-ab6e-31736449f4e9.webm)

# Screenshots

![image](https://user-images.githubusercontent.com/175305/235287343-d30b7025-8e1a-4de0-981c-7e6278ec34fd.png)

![image](https://user-images.githubusercontent.com/175305/235287348-9f9030f8-a844-4a28-88ca-d6c289963406.png)

![image](https://user-images.githubusercontent.com/175305/235287477-1a57cf19-d326-4409-a906-a150e272830b.png)

kfatehi commented:
Converted it to a draft because I forgot to implement the prosody features, which Riva supports, so it's just a matter of piping them through the proxy.

But would this PR even be approved? I think it's important to support Riva considering it's a fully offline, yet very high quality solution. One just need to have the GPU power to run the models.

kfatehi commented:
I would also want to reduce the blind code-copying I did of the IBM engine, prior to this being merged, but I wanted to get it out there because it does work and I want to see if there is interest for me to complete it to a higher standard of quality. As it stands now it works for me though.

ken107 commented:
I'll merge it.  This is indeed very cool.  Even though most users right now won't have the hardware needed to run it locally, the idea of having an offline next-gen TTS voice that's free to use is the future we'd like to get to.  When the time comes, perhaps the TTS subsystem can be deployed as a native app and extensions communicate with it via native messaging. (For reference, https://docs.nvidia.com/deeplearning/riva/user-guide/docs/tutorials-tts-contents.html)

ken107 commented:
Well, at this time I think only a few users will be able to use this feature, so it's basically experimental, it doesn't have to be perfect, as long as it doesn't affect existing functionalities. So you don't need to implement that prosody feature, or leave it for later it's fine. And using a proxy is fine too, it'll be a work in progress.

Native messaging is a way for extension to talk to a native app that's installed separately in the OS. The native app still has to do the gRPC proxying, so I think using HTTP like you're doing is no difference. So never mind about native messaging.

As long as someone with the necessary hardware can set it up, this will be a fun experimental function they can try out.

kfatehi commented:
@ken107 Ok great, it's ready for your review.

1. I implemented prosody (pitch and rate interpreted within Riva) which meant passing a rate of 1 to the player, and letting the proxy interpret it into Riva's preferred scale. Likewise pitch is rescaled properly to that which Riva wants. Works great.
2. Implemented streaming in the proxy (text -> wav -> ogg) so the latency is as low as possible (at the cost of lesser throughput, but this is worth it since we're talking personal use). This is working perfectly and it's super fast now regardless of sentence length.
4. Implemented prefetching, which seems to work well and made the transition from paragraph to paragraph even faster.
5. Fixed a [bug](https://github.com/kfatehi/riva_tts_proxy/issues/1) where multi-sentence was breaking Riva, this was as simple as tokenizing the input paragraph into sentences.
6. Added MIT license
7. Published to Docker hub
8. Updated the README to reflect the new API (JSON -> OGG stream) and how to use it straight from the Docker hub.
9. Got the Docker image down to 141MB compressed (380MB according to docker images) by rewriting the Dockefile w/ alpine linux.

All in all, quite comprehensive! I'm more proud of the proxy now than before. A good Samaritan rich in GPU power could in theory deploy this, crank the WEB_CONCURRENCY environment variable way up, and serve to many people. In the meantime, people with modern consumer GPUs can also use it.

Great job on read-aloud -- it was surprisingly easy to pull this off and it's still not even Sunday yet :)
